src/realizations/catchment/Bmi_Module_Formulation.cpp

```diff
@@ -384,31 +386,75 @@ namespace realization {
             determine_model_time_offset();
 
             // Output variable subset and order, if present
+            std::vector<std::string> out_headers;//define empty vector for headers
+            std::vector<std::string> out_units;//define empty vector for units for new json structure
             auto out_var_it = properties.find(BMI_REALIZATION_CFG_PARAM_OPT__OUT_VARS);
             if (out_var_it != properties.end()) {
                 std::vector<geojson::JSONProperty> out_vars_json_list = out_var_it->second.as_list();
+                //Check if the first item is an object type or string type.
+                //string type: old format; object type: new format
+                if (out_vars_json_list.size() > 0){
+                    std::string item_type = get_propertytype_name(out_vars_json_list[0].get_type());
+                    if (item_type == "String"){
+                        set_realization_file_format(true);
+                    }
+                }
                 std::vector<std::string> out_vars(out_vars_json_list.size());
-                for (int i = 0; i < out_vars_json_list.size(); ++i) {
-                    out_vars[i] = out_vars_json_list[i].as_string();
+                if (is_realization_legacy_format()){
+                    for (int i = 0; i < out_vars_json_list.size(); ++i) {
+                        out_vars[i] = out_vars_json_list[i].as_string();
+                    }
+                }
+                else{
+                    out_headers.resize(out_vars_json_list.size()); //assumption: number of vars = number of headers
+                    out_units.resize(out_vars_json_list.size()); //assumption: number of vars = number of units
+                    for (int i = 0; i < out_vars_json_list.size(); ++i) {
+                        out_vars[i] = out_vars_json_list[i].at("name").as_string();
+                        if(out_vars_json_list[i].has_key("header")){
+                            //indicates that a valid header is provided
+                            out_headers[i] = out_vars_json_list[i].at("header").as_string();
+                        }
+                        else{
+                            //indicates that header is not provided. The error actually returns a string.
+                            //in such cases, we assign variable name to the header.
+                            out_headers[i] = out_vars[i];
+                            std::stringstream ss;
+                            ss << "Header not provided for " << out_vars[i] << ". Using the variable name as header." << std::endl;
+                            LOG(ss.str(), LogLevel::WARNING); ss.str("");
+                        }
+                        out_units[i] = out_vars_json_list[i].at("units").as_string();
```

Comment:
It looks like these values aren't being saved anywhere once they're read from the config.

Also, we should parse them with UDUNITS2 here, and report an error if that fails, rather than waiting to try a conversion.